source/secure_tunneling.c

```diff
@@ -529,142 +452,1400 @@ static int s_init_data_tunnel_pair(
+
+    aws_linked_list_push_front(&secure_tunnel->queued_operations, &operation->node);
+
+    s_reevaluate_service_task(secure_tunnel);
+}
+
+struct aws_secure_tunnel_submit_operation_task {
+    struct aws_task task;
+    struct aws_allocator *allocator;
+    struct aws_secure_tunnel *secure_tunnel;
+    struct aws_secure_tunnel_operation *operation;
+};
+
+static void s_secure_tunnel_submit_operation_task_fn(struct aws_task *task, void *arg, enum aws_task_status status) {
+    (void)task;
+
+    int completion_error_code = AWS_ERROR_IOTDEVICE_SECURE_TUNNELING_SECURE_TUNNEL_TERMINATED;
+    struct aws_secure_tunnel_submit_operation_task *submit_operation_task = arg;
+
+    /*
+     * Take a ref to the operation that represents the secure tunnel taking ownership
+     * If we subsequently reject it (task cancel), then the operation completion
+     * will undo this ref acquisition.
+     */
+    aws_secure_tunnel_operation_acquire(submit_operation_task->operation);
+
+    if (status != AWS_TASK_STATUS_RUN_READY) {
+        goto error;
+    }
+
+    /*
+     * If we're offline fail it immediately.
+     */
+    struct aws_secure_tunnel *secure_tunnel = submit_operation_task->secure_tunnel;
+    if (secure_tunnel->current_state != AWS_STS_CONNECTED) {
+        completion_error_code = AWS_ERROR_IOTDEVICE_SECURE_TUNNELING_OPERATION_FAILED_DUE_TO_DISCONNECTION;
+        goto error;
+    }
+
+    s_enqueue_operation_back(submit_operation_task->secure_tunnel, submit_operation_task->operation);
+
+    goto done;
+
+error:
+    s_complete_operation(NULL, submit_operation_task->operation, completion_error_code, NULL);
+
+done:
+    aws_secure_tunnel_operation_release(submit_operation_task->operation);
+    aws_secure_tunnel_release(submit_operation_task->secure_tunnel);
+
+    aws_mem_release(submit_operation_task->allocator, submit_operation_task);
+}
+
+static int s_submit_operation(struct aws_secure_tunnel *secure_tunnel, struct aws_secure_tunnel_operation *operation) {
+    struct aws_secure_tunnel_submit_operation_task *submit_task =
+        aws_mem_calloc(secure_tunnel->allocator, 1, sizeof(struct aws_secure_tunnel_submit_operation_task));
+    if (submit_task == NULL) {
+        return AWS_OP_ERR;
+    }
+
+    aws_task_init(
+        &submit_task->task, s_secure_tunnel_submit_operation_task_fn, submit_task, "SecureTunnelSubmitOperation");
+    submit_task->allocator = secure_tunnel->allocator;
+    submit_task->secure_tunnel = aws_secure_tunnel_acquire(secure_tunnel);
+    submit_task->operation = operation;
+
+    aws_event_loop_schedule_task_now(secure_tunnel->loop, &submit_task->task);
+
+    return AWS_OP_SUCCESS;
+}
+
+/*********************************************************************************************************************
+ * Service Timing
+ ********************************************************************************************************************/
+
+static uint64_t s_min_non_0_64(uint64_t a, uint64_t b) {
```

Comment:
If we use this in mqtt5 maybe it should become a shared API in common

Reply:
https://github.com/awslabs/aws-c-common/pull/983